src/Components/ResourcePicker/ResourcePicker.tsx

```diff
@@ -0,0 +1,649 @@
+                            selectedOption.text
+                    );
+                }
+            }
+            return options;
+        },
+        []
+    );
+
+    const optionsFromResources = () => {
+        const newOptions: Array<IComboBoxOption> = [];
+        let lastHeader = '';
+        // after fetching resources, first start the dropdown with that
+        resourcesRef.current?.forEach((r) => {
+            if (r.subscriptionName && lastHeader !== r.subscriptionName) {
+                newOptions.push({
+                    key: r.subscriptionName,
+                    text: r.subscriptionName,
+                    itemType: SelectableOptionMenuItemType.Header
+                });
+                newOptions.push({
+                    key: r.id,
+                    text:
+                        getDisplayFieldValue(r) ||
+                        t('resourcesPicker.displayFieldNotFound', {
+                            displayField:
+                                AzureResourceDisplayFields[displayField],
+                            id: r.id
+                        }),
+                    data: r,
+                    styles: comboBoxOptionStyles
+                } as IComboBoxOption);
+                lastHeader = r.subscriptionName;
+            } else {
+                newOptions.push({
+                    key: r.id,
+                    text:
+                        getDisplayFieldValue(r) ||
+                        t('resourcesPicker.displayFieldNotFound', {
+                            displayField:
+                                AzureResourceDisplayFields[displayField],
+                            id: r.id
+                        }),
+                    data: r,
+                    styles: comboBoxOptionStyles
+                } as IComboBoxOption);
+            }
+        });
+        return newOptions;
+    };
+
+    // reset the dropdown options after fetching data, first options are resources, last items are the freeform added ones
+    useEffect(() => {
+        let newOptions: Array<IComboBoxOption> = [];
+        if (resourcesState.adapterResult.getCatastrophicError()) {
+            resourcesRef.current = null;
+            setError(resourcesState.adapterResult.getCatastrophicError());
+            newOptions = additionalOptions || [];
+            if (
+                selectedOption &&
+                !additionalOptions?.find((o) => o.text === selectedOption.text)
+            ) {
+                newOptions = newOptions.concat([selectedOption]);
+            }
+        } else if (!resourcesState.adapterResult.hasNoData()) {
+            const resources: Array<IAzureResource> =
+                resourcesState.adapterResult.result?.data;
+            resourcesRef.current = resources;
+
+            if (onResourcesLoaded) {
+                onResourcesLoaded(resources);
+            }
+
+            newOptions = mergeOptionsWithProps(
+                optionsFromResources(),
```

Comment:
can we please pass in the data to this function taht it is going to use? Setting the data to a ref a few lines before calling a function and then using that internally feels pretty dirty. If you need the ref for some other reason, at least pass in the value of the ref instead of using it internally. It will make it much more clear what dependencies this function has.